File: Algorithms/Math/0029_divide_two_numbers.py
#math
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        # handle overflow
        if dividend == -2147483648 and divisor == -1:
            return 2147483647
        
        # Determine the sign
        negative = (dividend < 0) != (divisor < 0)
        
        a = abs(dividend)
        b = abs(divisor)
        res = 0
        
        while a >= b:
            temp_divisor = b
            multiple = 1
            # shifting bits to the left to stimulate multiply by 2 
            while a >= (temp_divisor << 1):
                temp_divisor <<= 1
                multiple <<= 1
            
            # Subtract the largest chunk we found
            a -= temp_divisor
            res += multiple
            
        return -res if negative else res

File: Algorithms/Math/0029_divide_two_numbers.c
// Practice for: 0029_divide_two_numbers.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int divide(int dividend, int divisor) {
    if (dividend = INT_MIN && divisor == -1){
        return INT_MAX;
    }

    //determine sign (XOR logic)
    int negative = (dividend < 0) ^ (divisor < 0);
    long long a = abs(dividend);
    long long b = abs(divisor);
    long long res = 0;

    while (a >= b){
        long long temp_divisor = b;
        long long multiple = 0; //2^0
        while (a >= (temp_divisor << 1)){ //boundary checking, we check if it will overflow/greater should we bitshift
            temp_divisor << 1;
            multiple << 1;
        }
            a -= temp_divisor;
            res += multiple;
    }
    return negative ? (int)-res : (int)res;
}
